### ISSUE TITLE
Import/Export: Add all files to Export

### ISSUE THREAD

**Issue by mariobehling** [fossasia/open-event-orga-server#1217]
Related to Import Issue: https://github.com/fossasia/open-event-orga-server/issues/1216

Currently we only export the JSON files. Please also implement the export of:
* [x] all image files
* [x] all audio files
* [ ] add log activity in /admin/logs

**Comment by the-dagger** [comment]
Shouldn't be too hard @mariobehling. 
But we won't be needing audio files for the Android app. 
Also could you let me know what you mean by logs? 
Cheers.

**Comment by mariobehling** [comment]
@the-dagger Ok, agree. Logs is related to the admin area. Please look at the online installation and check it out yourself at ` /admin/logs` (as mentioned in the issue).

**Comment by aviaryan** [comment]
Since the need is that the exported zip can contain both direct links to resources on our server and embededded media files in the zip, I suggest we allow a config to be sent with an event export request. 

```js
{
"images": True,
"audio": False,
"video": False,
"slides": False
}
```

- From the above setting, the exported zip will have image links as relative and all other links (audio, video) as absolute. Likewise only images will be included in the zip. 
- If all the keys in the above setting are set to True, then everything will be included in the zip and all links will be relative. 

The android app/user can set the preferences according to their need and the zip will be generated accordingly. 
The Orga-server UI can show a form with checkboxes upon starting export of an event and so allow the user to select their need.

**Comment by mariobehling** [comment]
> This is ofcourse believing that someone will not require both absolute links to a media object and the media object itself at the same time. 

@aviaryan Let's think of the use case, that an app would come with all the data, but would update later version, e.g. updated speaker images or updated/additional audio.

How should we handle this? Should we define the API endpoints as well?

**Comment by aviaryan** [comment]
@mariobehling That is an interesting question. 
In other issue #1702, we have been discussing about versioning on the server so that the client apps can download only the data that is updated.
There we are describing versions of APIs like `speakers_ver`, `sessions_ver` and so. 
If the version num on server is more/different than one on the client, then fetch that API.
This works for the text data part but the problem is with images which need to be locally stored. 

How to know which images have been updated so that only they can be download ?
The best way for this is to check the change in URLs. But as the storage schema is currently set to use a fixed url for a fixed resource (e.g. event 1 speaker 42 photo = `/events/1/speakers/42/photo`) so the url will not change. 
So the problem goes back to storage schema issue (#1050). There I think we all agreed on using secret key so that paths can be made un-predictable. Please refer to my comment  https://github.com/fossasia/open-event-orga-server/issues/1050#issuecomment-231822943
Now what if we use `events/1/sessions/1/audio/<HASH>` instead of  `events/1/sessions/1/audio/<ENCRYPTED>`. 
*(HASH is the sha256 or md5 of the file stored there)*
The above method will still make the urls unpredictible and will also update the url in-case the object is updated. So this objective will also be fulfilled. 

( *As far as keeping distinct data on the server is concerned, it will get tricky but it will be manageable.* )

To sum it up, when the zip is first imported to generate the app, the app will store absolute links to the object. 
Next when the version info is updated and some APIs are re-fetched, it will compare the links and only download links which are changed. 

@mariobehling @the-dagger @mananwason 
Can the app-generator use the absolute links from the json and download the image files itself rather than relying on orga-server for them ? 
If no, then the only problem that remains is how to give both relative links to images and absolute links to images in the same json. 
There is the way to add two keys: 'logo' and 'logo_absolute' that will store the relative and absolute links respectively but I don't like it as it does not reflect the APIs.

**Comment by mariobehling** [comment]
@aviaryan Ok, I see. For the moment the priority is to make it work. Event though there are better options, we could also open issues and implement those options later.

The question is, if we get the data from the zip, then how does the app know where to fetch API data from? We need to add the info to the API endpoints into the JSON files. Or am I missing something?

**Comment by the-dagger** [comment]
@aviaryan If you are asking if the app can download the images if their link if provided in the json file, like  `http://2016.fossasia.org/img/fossasia-dark.png`, then yeah we can do that.

**Comment by aviaryan** [comment]
> The question is, if we get the data from the zip, then how does the app know where to fetch API data from? We need to add the info to the API endpoints into the JSON files. Or am I missing something?

Ok I guess we will have to add a meta.json to contain information like `root_url`. 
Once the app has root_url, then the API endpoints can be constructed. 
Eg root_url is opev.com, then api to check event data is at opev.com/api/v2/events/ID (ID is present in the json).

> @aviaryan If you are asking if the app can download the images if their link if provided in the json file, like http://2016.fossasia.org/img/fossasia-dark.png, then yeah we can do that.

@the-dagger To be precise, I was asking if the app-generator can do that or not ?

**Comment by mariobehling** [comment]
@aviaryan Yes, the app can do that.

**Comment by niranjan94** [comment]
I think @aviaryan's suggestions make perfect sense. 

Majorly based on that, we could have the flow like:

1. The generated zip will contain all the media files and paths relative to the media files inside the zip. The generated zip will also contain a `meta.json` file that will have the endpoint information. (In future we can use the file for more meta info).
1. The android app will be generated using this exported zip. The android app will store the media files and the relative path to the media files. (**Not url**). 
1. The URL to the media files on the server could be of the form `events/1/sessions/1/audio/<HASH>` where `<HASH>` is a **salted** sha256 hash of a string like `<session_id>_audio`. The salt pre-shared with the android app. So, the android app need not store the URL to media files anywhere. The URL can always be generated. A salted hash of a string like `<session_id>_audio` makes the URL unpredictable at the same time in makes sure the URL remains the even if the file data changes. It is also easily generatable by the android app. 
1. Now, the android can display the local images at first (specifying it as a placeholder in picasso or glide or whichever image caching library that is being used), and specify the source as the generated media URL. If the URL request fails, (like when there is no internet), the local images will anyway be already displayed. If the request succeeds, the image will be swapped. (if the image is the same as the local image, the user will see no change, else, the new image will be displayed and cached)

So, in this way, 

- We do not have to worry about image versions. 
- Android app will only store the local paths. 
- The URL to the media is always generated. (fast and easy to generate a salted hash). 
- We will still be keeping distinct media data on the storage server, since the URL never changes

**Note:**
- Encryption of a string like `<session_id>_audio` with the key pre-shared with the app can also be used instead of a salted-hash. 
- If the exported zip has URLs to media instead of local media and path, the android app generator can download the images and store the local path. The rest of the points (from 2) remain unchanged.

@mariobehling @the-dagger @aviaryan what are your opinions on the above ?

**Comment by mariobehling** [comment]
@niranjan94 Thanks for going through this as well.
One small change to the export. As a matter of fact exporting audio (and possibly video files) might be very resource intensive. Some people might not even use it. So, instead of exporting always everything, please give users the option to export e.g. image in the file and audio as a link only. I have added an issue about this here: https://github.com/fossasia/open-event-orga-server/issues/1739
@aviaryan Please add additional issues, e.g. to include a meta.json, in the tracker.
Thank you everyone!

**Comment by the-dagger** [comment]
@niranjan94 Looks fine to me as well.
But the thing is that we are not packaging the images with the android app, as it might increase the size of the app significantly.
Also we will need to hardcode the path for each image resource in our android app separately if we plan to package it inside the app.
So what we planned was to fetch the image on first app launch with network connection and then cache it for later use.

**Comment by aviaryan** [comment]
> So, instead of exporting always everything, please give users the option to export e.g. image in the file and audio as a link only. I have added an issue about this here: #1739

Yes I am making it so such that users can choose what type of media files they want in the zip before starting export. Here is how it looks right now. 
![export_options](https://cloud.githubusercontent.com/assets/4047597/16959313/19ea7a58-4e02-11e6-8a19-55e2190a1747.png)

> @aviaryan Please add additional issues, e.g. to include a meta.json, in the tracker.

Done #1745

**Comment by mariobehling** [comment]
@aviaryan Very good. 
* Please add a page in the tab and keep the "export event" as lightbox as shown here.
* On click "start" close the lightbox and show a progress bar on the page. 
* When "export" is finished show the export link on the page
Please follow up on this in the issue https://github.com/fossasia/open-event-orga-server/issues/1739
